lightning/src/ln/wire.rs

```diff
@@ -22,6 +22,24 @@ use io;
 use ln::msgs;
 use util::ser::{Readable, Writeable, Writer};
 
+/// Trait to be implemented by custom message (unrelated to the channel/gossip LN layers)
+/// decoders.
+pub trait Read {
+	/// The type of the message decoded by the implementation.
+	type CustomMessageType;
+	/// Decodes a custom message to `CustomMessageType`. If the given message type is known to the implementation and
+	/// the message could be decoded, must return `Ok(Some(message))`. If the message type
+	/// is unknown to the implementation, must return `Ok(None)`. If a decoding error
+	/// occur, must return `Err(DecodeError::X)` where `X` details the encountered error.
+	fn read<R: io::Read>(&self, message_type: u16, buffer: &mut R) -> Result<Option<Self::CustomMessageType>, msgs::DecodeError>;
+}
+
+// Really needs a better name
+pub(crate) enum MessageOrCustom<T> {
```

Comment:
So I think Jeff was originally thinking this would be in `PeerManager` with the original read code from https://git.bitcoin.ninja/index.cgi?p=rust-lightning;a=commitdiff;h=7a586455c10d835a425fa1de8ec447acbf9c67f3 there. That said, your approach here definitely cleans up the `PeerManager` message handling pipeline a lot. Still, the line count here is huge, I wonder if its easier to to just make `wire::Message` `pub(crate)` instead of `pub`, then just inline the `Custom` variant there instead of having a separate enum.

That way we'd avoid most of the diff, but still keep the read stuff nice and tight.

Reply:
Thanks that works as well indeed, I had a misunderstood I guess.